### DISCUSSION TITLE
Question: Is there any example of calling a golang function with a String argument and getting a String result back?

### DISCUSSION THREAD

**Discussion by lestephane** [jnr/jnr-ffi#224] [Q&A]
After having found a reference to this project in [this blog post](https://blog.dogan.io/2015/08/15/java-jni-jnr-go/), I've been trying to talk to a GO compiled shared library using 'simple' strings, but I'm getting `java.lang.UnsatisfiedLinkError: unknown`, and I don't know how to turn on more debugging information.

Java Interface

```Kotlin
interface HelloInterface {
  fun SayHello(who: String): String
}
```

Go file compiled with `-buildmode=c-shared` to a `libhello-linux-amd64.so` file

```Go
package main

import "fmt"

//export SayHello
func SayHello(who string) string {
  return fmt.Sprintf("Hello, %s!", who)
}

func main() {
  println(SayHello("World"))
}
```

Loader class
```Kotlin
object HelloInterfaceLoader {
    internal lateinit var INSTANCE: HelloInterface
    init {
        val props = System.getProperties() as Map<String, String>
        val os = props["os.name"]!!.toLowerCase(Locale.ENGLISH)
        val arch = props["os.arch"]!!.toLowerCase(Locale.ENGLISH)
        val libHelloName = "hello-$os-$arch"
        INSTANCE = LibraryLoader.create(HelloInterface::class.java).load(libHelloName)
    }
}
```
When I call the loader
```
fun main() {
    println(HelloInterfaceLoader.INSTANCE.SayHello("World"))
}
```

I get the error trace
```
Exception in thread "main" java.lang.UnsatisfiedLinkError: unknown
	at jnr.ffi.provider.jffi.AsmRuntime.newUnsatisifiedLinkError(AsmRuntime.java:40)
	at hello.HelloInterface$jnr$ffi$0.SayHello(Unknown Source)
	at hello.RunHelloKt.main(RunHello.kt:6)
	at hello.RunHelloKt.main(RunHello.kt)
```

Am I trying to do something that is simply not possible? If someone has managed to pass a string back and forth between java and GO using jnr-ffi, I'd love to get a working example. This is with Go 1.15.8. 

Cheers

**Comment by headius** [comment]
I am not familiar with how Go exports functions for C-like consumers. That would be my first question: is the function being exported from the library in a C-compatible way?

If you can turn this into a small repository that I can build and run I may be able to help you better.

**Comment by lestephane** [comment]
@headius That's the problem I tried to raise with this issue, because I don't have an example, I don't know how to go about giving you an example.

**Comment by headius** [comment]
@lestephane I was just hoping you might put together a small repository that has these files and builds them so I can at least get the same native Go binary that you are trying to use. Your interface and loader code are clearly not just Java... I think they are Scala code? Really I just need to see an example project that builds these pieces and then I can try to help you make it work the way you expect.

**Comment by headius** [comment]
I have moved this issue to a discussion.

**Reply by lestephane** [reply]
That makes senses, because I don't know when I'll be able to provide you with an answer. The moment has come and gone, I eventually used JNI shortly after filing the issue. But I wanted it to be on the record that an example of how to do this might be useful for the jnr-ffi to gain even more attention...
